chainerrl/agents/ppo.py

```diff
@@ -0,0 +1,265 @@
+        average_v_decay (float): Decay rate of average V, only used for
+            recording statistics
+        average_loss_decay (float): Decay rate of average loss, only used for
+            recording statistics
+    """
+
+    saved_attributes = ['model', 'optimizer']
+
+    def __init__(self, model, optimizer,
+                 gpu=None,
+                 gamma=0.99,
+                 lambd=0.95,
+                 phi=lambda x: x,
+                 value_func_coeff=1.0,
+                 entropy_coeff=0.01,
+                 update_interval=2048,
+                 minibatch_size=64,
+                 epochs=10,
+                 clip_eps=0.2,
+                 clip_eps_vf=0.2,
+                 average_v_decay=0.999, average_loss_decay=0.99,
+                 ):
+        self.model = model
+
+        if gpu is not None and gpu >= 0:
+            cuda.get_device_from_id(gpu).use()
+            self.model.to_gpu(device=gpu)
+
+        self.optimizer = optimizer
+        self.gamma = gamma
+        self.lambd = lambd
+        self.phi = phi
+        self.value_func_coeff = value_func_coeff
+        self.entropy_coeff = entropy_coeff
+        self.update_interval = update_interval
+        self.minibatch_size = minibatch_size
+        self.epochs = epochs
+        self.clip_eps = clip_eps
+        self.clip_eps_vf = clip_eps_vf
+
+        self.average_v = 0
+        self.average_v_decay = average_v_decay
+        self.average_loss_policy = 0
+        self.average_loss_value_func = 0
+        self.average_loss_entropy = 0
+        self.average_loss_decay = average_loss_decay
+
+        self.xp = self.model.xp
+        self.target_model = None
+        self.last_state = None
+
+        self.memory = []
+        self.last_episode = []
+
+    def _act(self, state, train):
+        xp = self.xp
+        with chainer.using_config('train', train):
+            b_state = batch_states([state], xp, self.phi)
+            action_distrib, v = self.model(b_state)
+            action = action_distrib.sample()
+            return cuda.to_cpu(action.data)[0], v.data[0]
+
+    def _train(self):
+        if len(self.memory) + len(self.last_episode) >= self.update_interval:
+            self._flush_last_episode()
+            self.update()
+            self.memory = []
+
+    def _flush_last_episode(self):
+        if self.last_episode:
+            self._compute_teacher()
+            self.memory.extend(self.last_episode)
+            self.last_episode = []
+
+    def _compute_teacher(self):
+        """Estimate state values and advantages of self.last_episode
+
+        TD(lambda) estimation
+        """
+
+        adv = 0.0
+        for transition in reversed(self.last_episode):
+            td_err = (
+                transition['reward']
+                + (self.gamma * transition['nonterminal']
+                   * transition['next_v_pred'])
+                - transition['v_pred']
+                )
+            adv = td_err + self.lambd * adv
```

Comment:
This line should be `adv = td_err + self.gamma * self.lambd * adv`.
See https://github.com/openai/baselines/blob/master/baselines/pposgd/pposgd_simple.py#L77